liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/server/DevMojo.java

```diff
@@ -594,18 +594,38 @@ private void addArtifacts(org.eclipse.aether.graph.DependencyNode root, List<Fil
         List<Artifact> updatedArtifacts = new ArrayList<Artifact>();
         for (Dependency dep : dependencies) {
             boolean newDependency = true;
+
+            // match dependencies based on artifactId, groupId
             for (Dependency existingDep : existingDependencies) {
-                if (dep.getArtifactId().equals(existingDep.getArtifactId())) {
-                    newDependency = false;
-                    break;
+                if (dep.getArtifactId().equals(existingDep.getArtifactId())
+                        && dep.getGroupId().equals(existingDep.getGroupId())) {
+
+                    // check for matching version as long as version is not null in existing
+                    // dependency or new dependency
+                    if ((dep.getVersion() != null && dep.getVersion().equals(existingDep.getVersion()))
+                            || (dep.getVersion() == null)) {
+                        if ((dep.getType() != null && dep.getType().equals(existingDep.getType())
+                                || (dep.getType() == null && existingDep.getType() == null))) {
+
+                            // check for matching scope or default "compile" scope
+                            if ((dep.getScope() != null && dep.getScope().equals(existingDep.getScope())
+                                    || (dep.getScope() == null && (existingDep.getScope() == null
+                                            || existingDep.getScope().equals("compile"))))) {
+                                newDependency = false;
+                                break;
+                            }
+                        }
+                    }
                 }
             }
+
             if (newDependency) {
                 log.debug("New dependency found: " + dep.getArtifactId());
                 try {
-                    Artifact artifact = getArtifact(dep.getGroupId(), dep.getArtifactId(), dep.getType(), dep.getVersion());
+                    Artifact artifact = getArtifact(dep.getGroupId(), dep.getArtifactId(), dep.getType(),
+                            dep.getVersion());
                     updatedArtifacts.add(artifact);
-                } catch (MojoExecutionException e) {
+                } catch (MojoExecutionException | IllegalArgumentException e) {
```

Comment:
What situation would throw IllegalArgumentException?

Reply:
Catches the case where `artifactId` or `groupId` is null, will throw an IllegalArgumentException.
```
[WARNING] groupId can neither be null, empty nor blank
```

Reply:
In that case with IllegalArgumentException, we should print a more useful message to state which dependency it is referring to, to help the user figure out where the problem is in their pom.